### PR TITLE
Implementation of BFN for discretized MNIST dataset

### PR DESCRIPTION
**Pull request by DvdGhl78**
Tried to implement a version of your BFN model for 2D data of discretized MNIST. It all works (training and processing), the only issue is that sample generation doesn't generate images that can be associated to MNIST numbers. Have you got any ideas why this happens? (maybe the network is too small?)

### CODE REVIEW

**Comment by Algomancer** [comment]
If you get this working, i'd be happy to merge it.

**Comment by matekrk** [comment]
Hi, I sort of struggle to generate mnist images as well. Looking at the training loss it seems that the training is reasonable as well, and then I obtain something similar. (I haven't used as sophisticated architectures as unets.)
sampled_9
sampled_4
bin_sampled_6
bin_sampled_5
bin_16_0 01_5000_sampled_3
bin_16_0 01_5000_sampled_0

Any ideas what should be the backbone for MNIST/CIFAR and/or how to tune the parameters for BFN? Thanks!

**Comment by angusturner** [comment]
Suspect this is probably a matter of the back-bone arch. I think a small conv or u-net would make a big difference. Will let you know if I get a chance to test this.

**Comment by DvdGhl78** [comment]
Tried with a much larger model (UNet with ~11M params) and also added weight EMA for inference. I still can't get decent results. Here's something:
**1000** steps:
1000_ema

**5000** steps:
5000_ema

Nonetheless the loss is very low and smooth (here's a sample, I've also trained on much more steps):
loss

**Comment by matekrk** [comment]
Hi, actually with larger U-net the generation works well, even after 100 epochs.
largeunet_bin_2_0 0001_100_dropout00_sampled_1
largeunet_bin_2_0 0001_100_dropout00_sampled_2
largeunet_bin_2_0 0001_100_dropout00_sampled_3
largeunet_bin_2_0 0001_100_dropout00_sampled_5
largeunet_bin_2_0 0001_100_dropout00_sampled_7
largeunet_bin_2_0 0001_100_dropout00_sampled_8

**Comment by DvdGhl78** [comment]
Thank you, I should only trained the model for a few more epochs. I'll fix the notebook and upload the correct version.

**Comment by thorinf** [comment]
https://github.com/thorinf/bayesian-flow-mnist

**Comment by DvdGhl78** [comment]
Ok should be fine.

**Comment by Algomancer** [comment]
lgmt
